[CALCITE-6322] Casts to DECIMAL types are ignored

mihaibudiu opened the pull request:
This PR addresses the problem that casts such as `CAST(X as DECIMAL(12, 2))` were actually interpreted as no-ops in Calcite.
With this change the value X is actually rounded to the specified precision and scale, and if the result overflows a runtime error is produced. This is in line with all the other SQL implementations I have checked.

This PR handles casts where the source type of X is an integer or a decimal type, but there may be additional cases to consider. We leave this for a later PR.

With this check quite a few tests would fail, so I had to insert additional casts to make them pass. This has changed the numeric precision of some results.

mihaibudiu commented:
I will add a commit handling the case of casting floating point values to decimals, it's important.

mihaibudiu commented:
This PR is ready for review. I consider this a major bug, since it essentially affects all DECIMAL computations.

mihaibudiu commented:
This PR is fixing a major calcite bug, which has been open for many years. It would be nice if someone could take a look.
Unfortunately, the changes affect many tests.
The problem is that casts to a DECIMAL(x, y) type were completely ignored; this PR implements them correctly.

snuyanzin commented on core/src/test/java/org/apache/calcite/test/InterpreterTest.java:
```diff
@@ -579,7 +579,7 @@ private static void assertRows(Interpreter interpreter,
         + "from (values ('a', -1.2), ('a', 2.3), ('a', 15)) as t(x, y)\n"
         + "group by x";
-    sql(sql).returnsRows("[a, -1.2, 15.0, 16.1, 5.366666666666667]");
+    sql(sql).returnsRows("[a, -1.2, 15.0, 16.1, 5.3]");
```
shouldn't in that case it be `5.4` ?
It seems rounding should be tweaked a bit
or did i miss anything?

mihaibudiu replied:
@snuyanzin I have written extensive comments about the rounding mode in Calcite in several places, e.g.
https://issues.apache.org/jira/browse/CALCITE-6322
I don't like this choice myself, but it's consistent in how Calcite handles rounding elsewhere.
If we want to change this we should first (1) write a spec about the semantics of casts in Calcite, (2) implement it consistently everywhere.

mihaibudiu replied:
I have no problem doing this work myself, but first everyone should agree on the desired result, because otherwise every piece of code in Calcite that handles conversions (and there are a lot of them) will do something different.

snuyanzin commented on babel/src/test/resources/sql/redshift.iq:
```diff
@@ -288,12 +288,12 @@ select percentile_disc(0.6) within group (order by sal) from emp group by deptno
 select stddev_pop(sal) from emp;
 EXPR$0
-1139.488618295281639802851714193820953369140625
+1139.48
```
it looks like here should be
`1139.49`

snuyanzin commented on babel/src/test/resources/sql/redshift.iq:
```diff
@@ -308,24 +308,24 @@ EXPR$0
+select var_samp(distinct CAST(sal AS DECIMAL(12, 4))) from emp;
 EXPR$0
-1512779.356060606
+1512779.3560
```
also  seems an issue with rounding

snuyanzin commented on babel/src/test/resources/sql/redshift.iq:
```diff
@@ -378,10 +378,10 @@ select empno, avg(sal) over (order by empno rows unbounded preceding) from emp w
 7521, 1425.00
-7654, 1366.666666666667
+7654, 1366.66
 7698, 1737.50
 7844, 1690.00
-7900, 1566.666666666667
+7900, 1566.66
```
same issue with rounding

snuyanzin commented on babel/src/test/resources/sql/redshift.iq:
```diff
@@ -523,78 +523,78 @@ select deptno, ratio_to_report(sal) over (partition by deptno) from emp;
-7900, 602.7713773341707792496890760958194732666015625
+7521, 141.4213
+7654, 585.9465
+7698, 585.9465
+7844, 602.7713
+7900, 602.7713
```
same here

zabetak commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -598,6 +598,13 @@ void testCastBooleanToNumeric(CastType castType, SqlOperatorFixture f) {
 
+  /** Test case for <a href="https://issues.apache.org/jira/projects/CALCITE/issues/CALCITE-6322">
+   *  [CALCITE-6322] Casts to DECIMAL types are ignored</a>. */
+  @Test void testDecimalDecimalCast() {
+    final SqlOperatorFixture f = fixture();
+    f.checkScalar("CAST(1.123 AS DECIMAL(4, 0))", "1", "DECIMAL(4, 0) NOT NULL");
+  }
+
```
Instead of adding a new method should we rather enrich `testCastToExactNumeric`?

zabetak commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -598,6 +598,13 @@ void testCastBooleanToNumeric(CastType castType, SqlOperatorFixture f) {
   }
 
+  /** Test case for <a href="https://issues.apache.org/jira/projects/CALCITE/issues/CALCITE-6322">
+   *  [CALCITE-6322] Casts to DECIMAL types are ignored</a>. */
```
The PR claims to add support for three kinds of CASTS:
* decimal to decimal
* int to decimal
* double/float to decimal

I've seen a few tests for the first category, do we have something for the other two? If not then I think we should add.

Moreover, I've noticed some tests that are currently disabled by the `DECIMAL` flag. A few of them should be fixed by this PR so please enable those that are relevant.

mihaibudiu replied:
I will re-enable many of these tests, but this doesn't completely solve the cases covered by SqlOperatorTest.DECIMAL. I have filed a new issue https://issues.apache.org/jira/browse/CALCITE-6379 that covers arithmetic. Moreover, the current PR also does not cover casts from STRING and INTERVAL. I have filed  https://issues.apache.org/jira/browse/CALCITE-6380 to cover these two cases.

zabetak reviewed (commented):
Overall the PR looks good, but the explicit CAST that were added make me a bit skeptical so I would like to understand better why they are needed. Apart from that I left some other small comments/suggestions.

mihaibudiu commented on babel/src/test/resources/sql/redshift.iq:
```diff
@@ -523,78 +523,78 @@ select deptno, ratio_to_report(sal) over (partition by deptno) from emp;
 
 # STDDEV_POP
-select empno, stddev_pop(comm) over (order by empno rows unbounded preceding) from emp where deptno = 30 order by 1;
+select empno, stddev_pop(CAST(comm AS DECIMAL(12, 4))) over (order by empno rows unbounded preceding) from emp where deptno = 30 order by 1;
```
Initially I had filed issue https://issues.apache.org/jira/browse/CALCITE-6324 about this problem. See the comments there as well. Indeed, in Calcite, as in Oracle, the result type of an aggregate is the same as the type of the value aggregated. 

Moreover, the Calcite implementation of aggregates also uses *internally* the same type for all calculations. Since STDDEV needs to square values, it requires double the precision not to overflow. So without changing the type here, this program would fail because of overflows for decimal values (prior to this PR such overflows were simply ignored, which is wrong).

The problem of internal overflows could be solved in two ways:
(1) change the internal implementation of all aggregation functions that could overflow to use a different precision for intermediate results. 
(2) require the user to use a type wide enough to accommodate possible overflows

Now, unfortunately the compiler cannot choose safely a type to guarantee that the results won't overflow, even for a simple aggregate such as AVG. The reason is that, although the compiler has a bound on the precision of the values aggregated, in general it has no bound on the number of rows in the table at runtime (especially for cases like incremental view maintenance, where the plans generated need to work even for future contents of the tables). So no matter what precision the compiler may choose for internal computations, this precision may be exceeded at runtime for some input data. 

This leaves the other alternative: the user should choose a precision high enough to carry all computations. Although the user may also not know the size of the tables, at least the user has some domain-specific knowledge. So this PR adopts the approach in (2), which is also compatible with the existing Calcite implementation of aggregates.

If we want to implement (1), it will require much more work, we have to modify the implementation of all Calcite aggregates, choosing an arbitrary type for intermediate results.

mihaibudiu replied:
Calcite does not specify anywhere the type of the results produced by an aggregate function. However, the type inference code for aggregates implies that the result type is the same as the data type. If this is correct, these results were wrong, because the scale of the results is not the same as the scale of the input data. 

As you see, many of these problems stem from the fact that there is no written spec for what Calcite should be doing.

I don't have access to an Oracle database to check. My knowledge of Oracle for aggregations comes from their published documentation, e.g.: https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/STDDEV.html

Postgres returns double if inputs are integers, or DECIMAL otherwise, a completely different behavior from Oracle or Calcite.

mihaibudiu replied:
I agree that this is unpleasant. However, please note that this implementation only affects the Enumerable-based implementation, which is used to evaluate the Quidem tests. Other runtimes can choose other strategies for the precision of intermediate results.

There was a discussion about this in JIRA: https://issues.apache.org/jira/browse/CALCITE-6324, where @julianhyde pointed out that the problems is indeed the strategy chosen for implementing functions such as STDDEV.

So I think this is really a separate issue. If we think that the implementation of STDDEV (and a few other aggregates in Calcite's runtime is wrong, we should file an issue about that and address it. Then we can revert these tests to use a narrower type for the result.

zabetak replied:
The Enumerable implementation is used in production for various systems so it is not a test-only thing that's why I raised some concerns.

I don't have something more to add to this discussion. We may not fully agree on everything but that is fine; I trust you to advance this further as you see fit.

NobiGo replied:
@zabetak Now, In Calcite **REGR_SXX and similar aggregate functions return the wrong data type**  The detail please read https://issues.apache.org/jira/browse/CALCITE-4924  and I try to use another way to make the return type more reasonable by calculate.  @mihaibudiu This issue looks same as **CAST a literal to DECIMAL type return wrong result** https://issues.apache.org/jira/projects/CALCITE/issues/CALCITE-4871. If we can't resolv CALCITE-4824, Then we can't handlt the 4871

mihaibudiu replied:
I think that the current bug fix would solve both CALCITE-4924 and CALCITE-4871.
The result type is correctly inferred, but the cast to decimal does nothing.

I will take a look to see whether it would be easy to change the functions that compute STDDEV to use a higher precision internally. But that should probably be a separate PR.

mihaibudiu commented on druid/src/test/java/org/apache/calcite/test/DruidAdapter2IT.java:
```diff
@@ -1909,10 +1909,14 @@ private void checkGroupBySingleSortLimit(boolean approx) {
+        + "  DruidQuery(table=[[foodmart, foodmart]],"
+        + " intervals=[[1900-01-09T00:00:00.000Z/2992-01-10T00:00:00.000Z]],"
+        + " projects=[[$63, $90, $91, $89]], groups=[{0}], aggs=[[SUM($1), SUM($2), SUM($3)]],"
+        + " post_projects=[[$0, /($1, $2), CASE(=($3, 0), CAST(1:DECIMAL(19, 0)):DECIMAL(19, 0),"
```
I suspect this is some lack of optimization in the Druid code generator, which doesn't remove redundant casts. If we think this is a bug, I suggest we file a new issue to deal with it.

zabetak replied:
I suppose this was addressed by other PRs? If that's the case you can mark this discussion as resolved.

mihaibudiu commented on core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:
```diff
@@ -1836,7 +1836,7 @@ public static boolean isValidDecimalValue(@Nullable BigDecimal value, RelDataTyp
       final int intDigits = value.precision() - value.scale();
       final int maxIntDigits = toType.getPrecision() - toType.getScale();
-      return intDigits <= maxIntDigits;
+      return (intDigits <= maxIntDigits) && (value.scale() <= toType.getScale());
```
The question is really about what this function is supposed to implement.
My assumption is that this function decides whether a BigDecimal value can be represented exactly in the specified type. In this case I believe that both conditions need to be satisfied. If the scale is not large enough, the number may need rounding to be represented, so the representation won't be exact.

zabetak replied:
If we don't know what it does and why it exists better not modify it. Like this we avoid accidental changes that are not directly related to this PR.

mihaibudiu replied:
I will check, this may be a leftover that is not needed here, but is needed for https://issues.apache.org/jira/browse/CALCITE-3522. I expect that eventually we will need this code.

mihaibudiu replied:
This change is actually necessary, so I will create a new helper function for it to avoid modifying the existing one.

mihaibudiu commented:
@zabetak I have tried to address your comments. In the process I found at least one more bug in my PR, the rounding mode is really DOWN and not FLOOR. This was found through the new tests you requested.

zabetak reviewed (commented):
Currently, the PR brings in some net improvements but also some potentially debatable breaking changes. Let's continue iterating on the discussion of the newly added CASTs till we reach consensus.

mihaibudiu commented on babel/src/test/resources/sql/redshift.iq:
```diff
@@ -523,78 +523,78 @@ select deptno, ratio_to_report(sal) over (partition by deptno) from emp;
-7521, 800.00
-7654, 2200.00
-7698, 2200.00
-7844, 2200.00
-7900, 2200.00
+7499, 300.0000
+7521, 800.0000
+7654, 2200.0000
+7698, 2200.0000
+7844, 2200.0000
+7900, 2200.0000
```
If I use `(CAST(comm as DECIMAL(12, 2))` the results will be unchanged. I will make that change to make the diff smaller.

The implementation of casts to decimal in Calcite treats them as noops, which is clearly wrong. This bug was masking many other potential problems, which now will be apparent. For example, aggregation results were also incorrectly computed. 

There is no clear spec in Calcite how aggregation results are computed (i.e., the precision of intermediate results, or the precision of the final results). SQL dialects differ in this respect, and the standard does not mandate a specific implementation. The current Calcite implementation is a reasonable one, but it won't work for this example without the manually-inserted cast, because overflows in decimal computations will cause the computation to fail. If we think that the aggregation implementation in Calcite is wrong, we should file a separate issue.

mihaibudiu commented:
@zabetak I think I have addressed your comments. I moved part of this PR into in a separate one which has been merged https://github.com/apache/calcite/pull/3779. So what's left should be less controversial. I will also submit a new PR for https://issues.apache.org/jira/browse/CALCITE-6380 which will complete the implementation of DECIMAL tests.

zabetak commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -1725,13 +1746,13 @@ void testCastToBoolean(CastType castType, SqlOperatorFixture f) {
-              + "when 2 then 4.543 else null end",
-          "DECIMAL(5, 3)", "4.543");
       f.checkScalarExact("case 1 when 1 then 11.2 "
               + "when 2 then 4.543 else null end",
-          "DECIMAL(5, 3)", "11.200");
+          "DECIMAL(5, 3)", "11.2");
```
This test is still under DECIMAL flag. It is not fixed?

mihaibudiu replied:
Yes, this one still doesn't work. I will have to investigate why.
Note that here I am only implementing casts, so if there is a bug in the implementation of CASE it won't be affected.

zabetak commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -11861,8 +11874,8 @@ private static void checkArrayConcatAggFuncFails(SqlOperatorFixture t) {
     f.checkScalarExact("ceil(cast(3.5 as double))", "DOUBLE NOT NULL", "4.0");
-    f.checkScalarExact("ceil(cast(3.45 as decimal))",
-        "DECIMAL(19, 0) NOT NULL", "4");
+    f.checkScalarExact("ceil(cast(3.45 as decimal(19, 1)))",
+        "DECIMAL(19, 1) NOT NULL", "4");
```
Why do we change the default type (`DECIMAL(19,0)`)?

zabetak reviewed (commented):
Please consider also adding an entry in history.md about the potential breaking changes in this PR and if necessary update the reference doc for the behavior of the CAST function.

mihaibudiu reviewed (commented):
@zabetak I apologize for the delay in replying to your comments, I was on vacation. I will submit a new commit addressing them.
